### PR TITLE
Fix #7606 "ld: framework not found CoreImage" Bug

### LINKED ISSUE #7606: opencv does not build on Mac OS X 10.10 with Xcode 7
##### System information (version)

- OpenCV => master
- Operating System / Platform => Mac OS X 10.10
- Compiler => Xcode 7 (clang)

##### Detailed description

opencv fails to build, because the `CoreImage` framework can not be found. That is strange, because I can actually see it in the SDK at `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/CoreImage.framework`.

However, as opencv links agains `QuartzCore.framework`, which already includes `CoreImage`, we do not actually need to link against `CoreImage`. Removing `CoreImage` leads to a successfull build.

##### Steps to reproduce

```
mkdir build
cd build

cmake -D CMAKE_BUILD_TYPE=RELEASE  -D CMAKE_INSTALL_PREFIX=/tmp/opencv  -D INSTALL_C_EXAMPLES=YES -D INSTALL_PYTHON_EXAMPLES=YES -D OPENCV_EXTRA_MODULES_PATH=`pwd`/../../opencv_contrib/modules -D BUILD_EXAMPLES=ON ..

make
```

will fail with

```
ld: framework not found CoreImage
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libopencv_videoio.3.1.0.dylib] Error 1
```

### PR DESCRIPTION
**Pull request by DabeDotCom**
This affects OpenCV 2.4 on older Macs with Xcode versions
before 7.0 (i.e., Mac OS X versions older than 10.11 El Capitan)

resolves #7606

### This pullrequest changes

For older versions of Xcode, the "CoreImage" framework is included with QuartzCore, and shouldn't be listed explicitly.  This appends "-framework CoreImage" to the highgui module only when XCODE_VERSION is greater than 7.0.

(FWIW — I had also considered testing against DARWIN_MAJOR_VERSION > 14, instead...  "Half of one, six-dozen of the other" ?)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -923,6 +924,9 @@ endif(DEFINED WITH_GIGEAPI)
 
 if(DEFINED APPLE)
   status("    AVFoundation:"   HAVE_AVFOUNDATION   THEN YES                                        ELSE NO)
+  if(WITH_COREIMAGE OR HAVE_COREIMAGE)
+  status("    CoreImage:"      HAVE_COREIMAGE      THEN "YES (deprecated)"                         ELSE NO)
```
Comment (alalek) [inline CMakeLists.txt]: Why "deprecated"?
Looks like it is enabled on modern environments: `XCODE_VERSION VERSION_GREATER 7.0`

**Comment by DabeDotCom** [comment]
Sorry, I must have overlooked that when copy-pasting (i.e., it was "a problem with the keyboard controller")

That comment is innocuous; I can remove it.

-- 
:- Dabe

File: CMakeLists.txt (hunk 1 above)
Comment (alalek) [inline CMakeLists.txt]: @DabeDotCom Yes, please remove "deprecated" comment.

**Comment by mshabunin** [comment]
I think variant with `XCODE_VERSION VERSION_GREATER 7.0` check in _highgui/CMakeLists.txt_ is better, because it does not add options which nobody will use.

And your comment can describe the reason of this check:
>For older versions of Xcode, the "CoreImage" framework is included with QuartzCore, and shouldn't be listed explicitly.

**Comment by alalek** [comment]
Agree with @mshabunin - extra variable is not required here.
@DabeDotCom Could you check patch #7916 into master branch too? (we haven't access to old MacOSX)

**Comment by alalek** [comment]
Probably it is better to remove CoreImage dependency at all. Looks like it is not used in OpenCV code.

**Comment by alalek** [comment]
I have updated this patch according the comment above.
:+1:

**Comment by pmalhaire** [comment]
I am using OSX 10.9.5

I had to do this patch in order to make the build work, unfortunately I don't know how (yet) how to do it without impacting upper versions of OSX

```diff
diff --git a/modules/videoio/CMakeLists.txt b/modules/videoio/CMakeLists.txt
index d079634..fb64894 100644
--- a/modules/videoio/CMakeLists.txt
+++ b/modules/videoio/CMakeLists.txt
@@ -181,7 +181,7 @@ if(HAVE_AVFOUNDATION)
     list(APPEND VIDEOIO_LIBRARIES "-framework AVFoundation" "-framework QuartzCore")
   else()
     list(APPEND videoio_srcs ${CMAKE_CURRENT_LIST_DIR}/src/cap_avfoundation_mac.mm)
-    list(APPEND VIDEOIO_LIBRARIES "-framework Cocoa" "-framework Accelerate" "-framework AVFoundation" "-framework CoreGraphics" "-framework CoreImage" "-framework CoreMedia" "-framework CoreVideo" "-framework QuartzCore")
+    list(APPEND VIDEOIO_LIBRARIES "-framework Cocoa" "-framework Accelerate" "-framework AVFoundation" "-framework CoreGraphics" "-framework CoreMedia" "-framework CoreVideo" "-framework QuartzCore")
   endif()
 endif()
```

**Pull request merged**